benchmarks/polys.py

```diff
@@ -1,17 +1,146 @@
+        D = -1 + self.x * prod((self.y)[:n])
+        f = D * (3 + self.x * prod((self.y)[:n]))
+        g = D * (-3 + self.x * prod((self.y)[:n]))
+        fp, gp = Poly(f, self.x, *(self.y)[:n]), Poly(g, self.x, *(self.y)[:n])
+        fpe, gpe = self.R.from_sympy(f), self.R.from_sympy(g)
+        return f, g, fp, gp, fpe, gpe
+
+
+class PolyPrem:
+    """Benchmark for the prem method"""
+
+    def setup(self, n):
+        self.x, *self.y = symbols("x, y1:9")
+        self.R1 = [self.x] + list(self.y)
+        self.R = ZZ[self.R1]
+
+        self.values = {}
+
+        (self.f_LinearDenseQuadraticGCD, self.g_LinearDenseQuadraticGCD,
+        self.fp_LinearDenseQuadraticGCD, self.gp_LinearDenseQuadraticGCD,
+        self.fpe_LinearDenseQuadraticGCD, self.gpe_LinearDenseQuadraticGCD) = LinearDenseQuadraticGCD().generate(n)
+
+        (self.f_SparseGCDHighDegree, self.g_SparseGCDHighDegree,
+        self.fp_SparseGCDHighDegree, self.gp_SparseGCDHighDegree,
+        self.fpe_SparseGCDHighDegree, self.gpe_SparseGCDHighDegree) = SparseGCDHighDegree().generate(n)
+
+        (self.f_QuadraticNonMonicGCD, self.g_QuadraticNonMonicGCD,
+        self.fp_QuadraticNonMonicGCD, self.gp_QuadraticNonMonicGCD,
+        self.fpe_QuadraticNonMonicGCD, self.gpe_QuadraticNonMonicGCD) = QuadraticNonMonicGCD().generate(n)
+
+        (self.f_SparseNonMonicQuadratic, self.g_SparseNonMonicQuadratic,
+        self.fp_SparseNonMonicQuadratic, self.gp_SparseNonMonicQuadratic,
+        self.fpe_SparseNonMonicQuadratic, self.gpe_SparseNonMonicQuadratic) = SparseNonMonicQuadratic().generate(n)
+
+
+class TimePolyPremFast(PolyPrem):
+
+    # Test for n=1, 3, 5, 8 with the fast methods
+    params = [1, 3, 5, 8]
+
+    def time_prem_SparseGCDHighDegree(self, n):
+        self.values['prem_SparseGCDHighDegree'] = prem(self.f_SparseGCDHighDegree, self.g_SparseGCDHighDegree, self.x)
+
+    def time_prem_SparseNonMonicQuadratic(self, n):
+        self.values['prem_SparseNonMonicQuadratic'] = prem(self.f_SparseNonMonicQuadratic, self.g_SparseNonMonicQuadratic, self.x)
+
+    def time_Polyprem_SparseGCDHighDegree(self, n):
+        self.values['Polyprem_SparseGCDHighDegree'] = self.fp_SparseGCDHighDegree.prem(self.gp_SparseGCDHighDegree)
+
+    def time_Polyprem_QuadraticNonMonicGCD(self, n):
+        self.values['Polyprem_QuadraticNonMonicGCD'] = self.fp_QuadraticNonMonicGCD.prem(self.gp_QuadraticNonMonicGCD)
+
+    def time_Polyprem_SparseNonMonicQuadratic(self, n):
+        self.values['Polyprem_SparseNonMonicQuadratic'] = self.fp_SparseNonMonicQuadratic.prem(self.gp_SparseNonMonicQuadratic)
+
+    def time_PolyElement_prem_SparseGCDHighDegree(self, n):
+        self.values['PolyElement_prem_SparseGCDHighDegree'] = self.fpe_SparseGCDHighDegree.prem(self.gpe_SparseGCDHighDegree)
+
+    def time_PolyElement_prem_SparseNonMonicQuadratic(self, n):
+        self.values['PolyElement_prem_SparseNonMonicQuadratic'] = self.fpe_SparseNonMonicQuadratic.prem(self.gpe_SparseNonMonicQuadratic)
```

Comment:
The point is to have different subclasses of e.g. `LinearDenseQuadraticGCD` then each subclass would just use
```python
    (self.f, self.g, self.fp, self.gp, self.fpe, self.gpe) = self.generate()
```
It could be something like this:
```python
class _PolyGCDExample:
    def setup(self, n):
        (self.f, self.g, self.fp, self.gp, self.fpe, self.gpe) = self.generate()

class _LinearDenseQuadraticGCD:
    def generate(self, n):
        ...
        return f, g, ...

class _TimePREM(_PolyGCDExample):

    def time_prem(self):
        self.values['prem'] = prem(self.f, self.g)

    def time_prem_PolyElement:
        self.values['prem_PolyElement'] = ...

    def teardown(self, n):
        # here there should be code to check that the answers in self.values are correct.

class TimePREMLinearDenseQuadraticGCD(_LinearDenseQuadraticGCD, _TimePREM):
    params = [1, 3, 5, 8]
```
Alternatively it could be like:
```python
def generate_linear_dense(n):
    ...

def generate_polys(case, n):
    if case == 'linear_dense':
        return generate_linear_dense(n)
    elif case ...

class TimePREM:
    params = [('linear_dense', 1), ('linear_dense', 2), ...]

    def setup(self, case, n):
        self.f, ... = generate_polys(case, n)

    def time_prem(self, case, n):
        ...

    def time_prem_PolyElement(...)
```
There are lots of ways to do this that don't involve the kind of repetitive code I have highlighted here.

We need a structure that makes it easy to add new benchmarks for different functions and different cases without large amounts of repetitive code.

Also different ways of structuring this lead to different output from asv which may be easier or harder to understand. Using multiple parameters leads to output like this:
```
[ 44.11%] ··· solve.TimeMatrixOperations.time_rref                            ok
[ 44.11%] ··· ======== =========== ============ ============
              --                       param2               
              -------- -------------------------------------
               param1       0           2            5      
              ======== =========== ============ ============
                 3      211±0.2μs   27.3±0.1ms   51.4±0.1ms 
                 4      413±0.7μs   68.4±0.2ms   78.6±0.3ms 
              ======== =========== ============ ============
```
If the parameters are not just numbers then that makes it easier to understand. I'm not sure f it is possible to tell `asv` to use something other than just `param1` or `param2` as the parameter names. String parameter values will show in the output though:
```
[ 44.75%] ··· solve.TimeMatrixSolve.time_solve_sym                            ok
[ 44.75%] ··· ======== =========
               param1           
              -------- ---------
                 GE     681±2ms 
                 LU     766±8μs 
                ADJ     202±3ms 
              ======== =========
```
Here we probably want the benchmarks to show how different methods compare for the same class of polynomials. Maybe we want something more like this:
```python
class _LinearDenseQuadraticGCD:
    def generate(self, n):
        ...
        return f, g, ...

class _TimePREM:

    params = [('prem', 1), ('prem', 2), ..., ('prem_PolyElement', 1), ...]

    def setup(self, method, n):
        f, g = self.generate(n)
        if method == 'prem':
            self.func = lambda: prem(f, g)
        elif ...

    def time(self, method, n):
        self.func()

class TimePREMLinearDense(_TimePREM, _LinearDenseQuadraticGCD):
    params = [('prem', 1), ('prem', 2), ..., ('prem_PolyElement', 1), ...]
```
Then for each class of polynomials we will see a table showing how each method scales with `n`.